src/include/platform/internal/GenericPlatformManagerImpl_FreeRTOS.cpp

```diff
@@ -68,33 +66,38 @@ CHIP_ERROR GenericPlatformManagerImpl_FreeRTOS<ImplClass>::_InitChipStack(void)
     if (mChipEventQueue == NULL)
     {
         ChipLogError(DeviceLayer, "Failed to allocate CHIP event queue");
-        ExitNow(err = CHIP_ERROR_NO_MEMORY);
+        return CHIP_ERROR_NO_MEMORY;
     }
 
-    // Call up to the base class _InitChipStack() to perform the bulk of the initialization.
-    err = GenericPlatformManagerImpl<ImplClass>::_InitChipStack();
-    SuccessOrExit(err);
+    return CHIP_NO_ERROR;
+}
 
-exit:
-    return err;
+template <class ImplClass>
+CHIP_ERROR GenericPlatformManagerImpl_FreeRTOS<ImplClass>::_InitChipStack(void)
+{
+    ReturnErrorOnFailure(_InitPlatformObjects());
+
+    // Call up to the base class _InitChipStack() to perform the bulk of the initialization.
+    ReturnErrorOnFailure(GenericPlatformManagerImpl<ImplClass>::_InitChipStack());
+    return CHIP_NO_ERROR;
 }
 
 template <class ImplClass>
 void GenericPlatformManagerImpl_FreeRTOS<ImplClass>::_LockChipStack(void)
 {
-    xSemaphoreTake(mChipStackLock, portMAX_DELAY);
+    xSemaphoreTakeRecursive(mChipStackLock, portMAX_DELAY);
```

Comment:
Why we need to make all lock recursive? this will make LockChipStack get called multiple times by the same process/thread without seeing a problem. I think this change will hide some potential Lock/Unlock misuse.

Reply:
No, this is not a misuse.
The SendMessage can be called in CHIP mainloop handler or outside the mainloop handler, in the first case, we already have the stack lock, but in the second case, we does not have the stack lock.

Reply:
I don't means you misuse this one. With current implementation, if we attempt to perform the "LockChipStack" operation if we have already locked the chip stack within the platform layer, it should fail.

After we change to recursive lock, we will not see this failure in this scenario. Typically, a recursive mutex tracks the number of times it has been locked, and requires equally many unlock operations to be performed before other threads may lock it. It is a powerful construct, but it also make us more difficult to debug the issue in case of failure.

Reply:
Recursive locking is a good way to make incorrect synchronization _appear_ to work while still being subtly broken. We shouldn't use them.